Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al

```diff
@@ -0,0 +1,332 @@
+    end;
+
+    local procedure VerifyOffsetIsCorrectForSelectedTimeZone(Offset: Duration; ExpectedOffsetInHours: Decimal)
+    var
+        ExpectedOffset: Duration;
+        OffsetMismatchErr: Label 'Offset did not match expected offset.', Locked = true;
+    begin
+        ExpectedOffset := ExpectedOffsetInHours * GetHourConversionFactor();
+        LibraryAssert.AreEqual(ExpectedOffset, Offset, OffsetMismatchErr);
+    end;
+
+    local procedure VerifyOffsetBetweenTimeZonesIsCorrect(Offset: Duration)
+    var
+        OffsetMismatchErr: Label 'Offset between time zones did not match expected offset.', Locked = true;
+        ExpectedOffset: Duration;
+    begin
+        ExpectedOffset := -9 * GetHourConversionFactor();
+        LibraryAssert.AreEqual(ExpectedOffset, Offset, OffsetMismatchErr);
+    end;
+
+    local procedure CheckWhetherTimeZoneSupportsDaylightSavingTime(TimeZoneId: Text): Boolean
+    begin
+        exit(DaylightSavingTimeInfo.TimeZoneSupportsDaylightSavingTime(TimeZoneId));
+    end;
+
+    local procedure VerifyDaylightSavingTimeIsNotSupported(SupportsDST: Boolean)
+    var
+        IncorrectDSTSupportErr: Label 'Time zone was not expected to support daylight saving time.', Locked = true;
+    begin
+        LibraryAssert.IsFalse(SupportsDST, IncorrectDSTSupportErr);
+    end;
+
+    local procedure VerifyDaylightSavingTimeIsSupported(SupportsDST: Boolean)
+    var
+        IncorrectDSTSupportErr: Label 'Time zone was expected to support daylight saving time.', Locked = true;
+    begin
+        LibraryAssert.IsTrue(SupportsDST, IncorrectDSTSupportErr);
+    end;
+
+    local procedure CheckWhetherDateTimeFallsWithinDSTForTheTimeZone(DateTimeToTest: DateTime; TimeZoneId: Text): Boolean
+    begin
+        exit(DaylightSavingTimeInfo.IsDaylightSavingTime(DateTimeToTest, TimeZoneId));
+    end;
+
+    local procedure VerifyDateTimeShowsAsBeingInDST(DateTimeIsInDST: Boolean)
+    var
+        IncorrectDSTSupportErr: Label 'DateTime was expected to show as being in daylight saving time.', Locked = true;
+    begin
+        LibraryAssert.IsTrue(DateTimeIsInDST, IncorrectDSTSupportErr);
+    end;
+
+    local procedure VerifyDateTimeShowsAsNotBeingInDST(DateTimeIsInDST: Boolean)
+    var
+        IncorrectDSTSupportErr: Label 'DateTime was not expected to show as being in daylight saving time.', Locked = true;
+    begin
+        LibraryAssert.IsFalse(DateTimeIsInDST, IncorrectDSTSupportErr);
+    end;
+
+    local procedure GetExpectedOffset(TimeZoneId: Text; InDaylightSavingPeriod: Boolean): Decimal
+    begin
+        case TimeZoneId of
+            GetTimeZoneIdForNonDSTTimeZone():
+                exit(-7);
+            GetTimeZoneIdForDSTTimeZone():
+                begin
+                    if InDaylightSavingPeriod then
+                        exit(2);
+                    exit(1);
+                end;
+        end;
+    end;
+
+    local procedure GetTimeZoneIdForNonDSTTimeZone(): Text
```

Comment:
If we've being honsest with ourselves, the name of this function should not be GetTimeZoneIdForNonDSTTimeZone(). It should rather be GetTimeZoneIdForUSMountainStandardTime. You are using a lot of hard coded values and "magic numbers" in your test - that could be improved.

Reply:
True, but the problem with this sort of functionality is that it does not exist yet. In a good test, you would have two ways of reaching the same result: the one you want to test and the control where you know the answer is correct.

The problem here is that there is no real alternative method to reach the correct offset; if there were, I would not have needed to create this module. The only thing I can do is use the same method using the TimeZoneMgt dotnet module, which would reliably use the same  time.

A possible alternative is using several time zones in the set, but I don't really see a way around using hard-coded information.

Reply:
I have made some changes to the test feature to introduce a wider array of time zones to test and a certain degree of randomness in order to make the test more robust.